themes/default/content/blog/csharp-static-analyzer/index.md

```diff
@@ -0,0 +1,132 @@
+
+Enter [PulumiCSharpAnalyzer](https://github.com/Zaid-Ajaj/pulumi-csharp-analyzer): a Roslyn-based code analyzer that adds compile-time checks to your Pulumi code and reports it as part of the build.
+
+Available today as a NuGet package, you can install the analyzer into your project as follows:
+
+```
+dotnet add package PulumiCSharpAnalyzer
+```
+
+Now your IDE, whether that is Rider, Visual Studio or VS Code with OmniSharp should report the error inside the IDE like any compile error or warning. Here is for example how it looks like inside Rider:
+
+![missing property Kind when initializing resource arguments](./missing-kind-property.png)
+
+Not just your IDE, it integrates into your build pipeline too. When you `dotnet build` your project, the detected errors will show up there:
+
+```bash
+> dotnet build
+Microsoft (R) Build Engine version 17.2.0+41abc5629 for .NET
+Copyright (C) Microsoft Corporation. All rights reserved.
+
+  Determining projects to restore...
+  All projects are up-to-date for restore.
+/Users/zaid/projects/AzureExample/Program.cs(11,63): warning MissingRequiredProperty: Missing required property Kind when initializing properties of type StorageAccountArgs [/Users/zaid/projects/AzureExample/AzureExample.csproj]
+  AzureExample -> /Users/zaid/projects/AzureExample/bin/Debug/net6.0/AzureExample.dll
+
+Build succeeded.
+```
+
+Besides detecting missing properties when creating resources, the analyzer also detects missing properties for function invokes which follow a similar pattern:
+
+![missing required parameter when invoking a SDK function](./error-missing-required-properties-function-invokes.png)
+
+Finally, an anti-pattern that the analyzer can detect is creating resources within an `.Apply(...)` call:
+
+![a warning being reported when creating a resource inside apply call](./resource-inside-apply.png)
+
+We always recommend developers to avoid creating resources inside `.Apply(...)` calls because the execution plan of Pulumi will be different during the `pulumi preview` and `pulumi up` phase.
+
+The reason is that when we have an instance of `Output<T>` from which we call `Apply`, the value of the instance is most likely _unknown_ during `pulumi preview` since it is only computed when Pulumi actually performs the deployment plan against APIs of cloud providers.
+
+## Limitations
+
+With the current features of the analyzer, it is already usable across all cloud provider SDKs. That said, there are limitations to the way the analyzer works. At the moment, when determining the missing properties, the logic only detects initialized properties when creating the object and diffs those against the required properties from the schema. This means that if you happened to use property _setters_ instead of property initializers, they are considered "not provided" and will report a warning for the following piece of code:
```

Comment:
```suggestion
With the current features of the analyzer, it is already usable across all cloud provider SDKs. That said, there are limitations to the way the analyzer works. At the moment, when determining the missing properties, the logic only detects initialized properties when creating the object and diffs those against the required properties from the schema. This means that if you happened to use property _setters_ instead of property initializers, they are considered "not provided" and will report a warning. For example, the following piece of code will produce a false-positive warning...
```